components/openid-token-builder/src/main/java/com/wso2telco/openidtokenbuilder/MIFEOpenIDTokenBuilder.java

```diff
@@ -153,23 +167,163 @@
      */
     public String buildIDToken(OAuthTokenReqMessageContext request,
                                OAuth2AccessTokenRespDTO tokenRespDTO) throws IdentityOAuth2Exception {
+        
+
         if (log.isDebugEnabled()) {
             log.debug("MSISDN : " + request.getAuthorizedUser().getUserName());
             log.debug("Generated access token : [" + tokenRespDTO.getAccessToken() + "]  for Authorization Code :  "
                     + request.getProperty("AuthorizationCode"));
         }
 
+        FederatedIdpMappingDTO fidpInstance = new FederatedIdpMappingDTO();
+
+        String accessTokenIssuedTime = getAccessTokenIssuedTime(tokenRespDTO.getAccessToken(), request);
+        String atHash = new String(Base64.encodeBase64(tokenRespDTO.getAccessToken().getBytes()));
+        String plainIDToken = null;
+        int curTime = (int) Calendar.getInstance().getTimeInMillis();
+
         OAuthServerConfiguration config = OAuthServerConfiguration.getInstance();
+        
         String issuer = config.getOpenIDConnectIDTokenIssuerIdentifier();
         int lifetime = Integer.parseInt(config.getOpenIDConnectIDTokenExpiration()) * 1000;
-        int curTime = (int) Calendar.getInstance().getTimeInMillis();
+
+        CustomIDTokenBuilder builder = new CustomIDTokenBuilder();
+        builder.setIssuer(issuer);
+        builder.setAudience(request.getOauth2AccessTokenReqDTO().getClientId());
+        builder.setAuthorizedParty(request.getOauth2AccessTokenReqDTO().getClientId());
+        builder.setExpiration(curTime + lifetime);
+        builder.setIssuedAt(curTime);
+        builder.setAtHash(atHash);
+        builder.setAuthTime(accessTokenIssuedTime);
+        boolean dataPublisherEnabled = mobileConnectConfig.getDataPublisher().isEnabled();
+        
+        try {
+
+            fidpInstance = dbConnection.retrieveFederatedAuthCodeMappings(request.getOauth2AccessTokenReqDTO()
+                    .getAuthorizationCode());
+        } catch (Exception e) {
+            log.error("Error while retrieving federated mapping information hence continuing regular flow"
+                    + request.getOauth2AccessTokenReqDTO().getAuthorizationCode());
+        }
+
+        if (fidpInstance.getFidpAuthCode() != null) {
+
+            log.info("Federated Identity Access Token Flow initiated");
+            
+            for (MobileConnectConfig.Provider prv : federatedIdps.getProvider()) {
+                federatedIdpMap.put(prv.getOperator(), prv);
+            }
+            
+            
+            try {
+
+                JSONObject fidpIDToken = tokenAuthenticationRequest(request, fidpInstance);
+
+                if (fidpIDToken.optString("access_token") != null) {
+
+                    dbConnection.insertFederatedTokenMappings(tokenRespDTO.getAccessToken(), request
+                            .getOauth2AccessTokenReqDTO().getAuthorizationCode(), fidpIDToken.get("access_token")
```

Comment:
Better to have the implementation in a different class or method as appropriate. having such will ease the unit test writing and better readability. Each set of code blocks supports one task need to be in methods with a meaningful name.